added browser back button listening

zcallan opened the pull request:
In response to #502, adds functionality of pushing a flag to the browser history state which closes the Dashboard modal when removed. Meaning, when the back button is pressed, the modal closes instead of the page changing.

arturi commented:
Hi! Thanks for the PR, will check it out soon.

goto-bus-stop commented:
Sweet! I think we should make this optional, in case it doesn't play nice with some existing client side router.

zcallan commented:
Thanks, sounds good. Thoughts on whether it should be enabled or disabled by default?

goto-bus-stop commented:
Hmm, Uppy should probably only do things to the history if the developer knows it will, so it might be best left off by default. I imagine debugging routing issues originating in the Dashboard would be difficult to track down if you're not aware that the Dashboard does that :)

arturi commented:
Yeah, I agree with @goto-bus-stop, leaning towards leaving it off by default.

zcallan commented:
Thanks guys, committed the changes 👍 

Also added it to the website to demonstrate it as an example in Dashboard if that's aight too!

zcallan commented:
Found a bug with double-pushing when manually closing and reopening the modal, but fixed it (1f9994c).

Only issue (well, somewhat of an issue) remaining is that if you open the modal, close it manually (not pressing back in the browser), then press back in the browser, the page won't change on the first click. You need 2 clicks to go back in browser history.

This is due to the fact that when `uppyDashboard: true` is pushed into the history state, it creates a new entry in the browser history. That entry cannot, to my current understanding, be removed with the current history API given by browsers.

Closest method is **replacing** the history state on modal close, but of course that doesn't clear the history entry in the browser, and the problem stands.

According to ([the MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/History_API#The_replaceState()_method)),

> `history.replaceState()` operates exactly like `history.pushState()` except that `replaceState()` modifies the current history entry instead of creating a new one. **Note that this doesn't prevent the creation of a new entry in the global browser history.**

Apart from that everything should be sweet. Thanks

zcallan commented:
I've made one more commit to somewhat remedy the above issue by simply using `history.go(-1)` to force the browser to go back one page when the modal is manually (for lack of a better word) closed.

The only thing I dislike about this approach is if something else is pushed into the browser history while the modal is open, and then the modal is manually closed while the other thing still exists in the history, the other thing would be removed and the `uppyDashboard: true` flag would (probably) stay in the history state.

Not so much an issue because it'd be such a rare occurrence, as well as the fact that the close on browser back button functionality is behind a flag in the Uppy settings, that it should be recognisable by the developer. It does still kind of irk me but I'm not sure if there's a better solution.

Thanks

goto-bus-stop commented:
> The only thing I dislike about this approach is if something else is pushed into the browser history while the modal is open, and then the modal is manually closed while the other thing still exists in the history, the other thing would be removed and the `uppyDashboard: true` flag would (probably) stay in the history state.

Can we use the `history.state` property to check that the `uppyDashboard: true` flag is the current entry, before doing `history.go(-1)`? https://devdocs.io/dom/history#history-state

Thanks for being so thorough! This is looking real good :)

zcallan commented:
Thanks @goto-bus-stop :)

Updated with a few more comments, replaced modal name with a variable and implemented change as ^above.

The issue we run into now is that if the modal is open while something else pushes to `history.state` and then the modal X close button is pressed, the other thing wouldn't be closed which is great but then when we press the browser back button, the other thing pushed to history state should be removed (which again is great and expected behaviour) but a second browser back button click removes the `uppyDashboard` history entry. Since the modal is already closed, that second click does nothing. A third browser back button click would work as expected.

Steps to reproduce:

1. Open dashboard modal with `browserBackButtonClose: true` flag set
2. Open devtools console, type `history.pushState({ test: true }, '')`
3. Close dashboard modal using the X close button in the UI
4. Type `history.state` to note that `{ test: true }` is the latest history entry
5. Press browser back button
6. Repeat step 4 but note that `{ uppy-Dashboard: true }` is now the latest history entry and page doesn't change (as it shouldn't - this part is expected behaviour ✅ )
7. Again press browser back button
8. Again repeat step 4 but now note that `history.state` is `null` and page still doesn't change (unexpected behaviour 🚫 )
9. Press browser back button, page should change (expected behaviour ✅ )

It's a somewhat convoluted process to recreate this bug as you can see.

As long as the developer doesn't push to `history.state` while the dashboard modal is open then this PR should work flawlessly.

The bug is not fatal or unrecoverable by any means - if anything, I would say the user will just suspect that they had misclicked the browser back button and simply press it again, overcoming the bug, but that's not perfect.

I can't really think of how we can overcome this issue - maybe some browser history god like  @mjackson knows a solution (hopefully he sees this 😂 ) - anyone any thoughts?

If not, maybe we should add a bold note in the docs under using the `browserBackButtonClose` flag  that you should be careful to not to push to the history state while the modal is open.

Thanks!

zcallan commented:
Fixed -- figured that if I didn't get rid of the event listener for `popstate` when the modal is closed, I could check if when the browser back button is pressed that the current browser history entry is the uppy dashboard modal and the modal is closed, we can just do `history.go(-1)` to remove the entry and fix the time travel.

Should be sweet to merge in, working flawlessly for me now.

arturi commented:
@zcallan hi, thanks for your work again! tried using this for the first time, when I click “forward” after going back, which closes Uppy modal, should it open the modal?

zcallan commented:
Thanks guys!

@arturi, correct it currently does not re-open the modal if you hit the "forward" button.

I would be cautious to add that functionality as I feel it would not be uncommon for developers to have some kind of additional setup when the modal is opened.

For instance, someone may want to add their own custom class on the `body` element when the modal is opened (alongside Uppy's own `uppy-Dashboard-isOpen` class).

If we invoke `this.openModal()` inside Uppy after the "forward" button has been pressed (and `uppy-Dashboard` has appeared in the history state again), we may be skipping any extra functionality a dev has added in response to the user trying to open the modal (like in the example instance).

A workaround for developers would be to have a `dashboard:open` event fire off when the modal is opened, and they can add their own extra custom functionality from inside their own handler.

I'm not sure if that itself should be a separate PR, then after that's ready and working we can add the ability for this PR (or create another new one) to be able to handle the "forward" button press more comfortably and less unassumingly.

Thoughts on that?

arturi commented:
Sorry it took so long, @zcallan, I’m truly very sorry! Something urgent always pushed this further.

Ideally, I think we’d like this to be another plugin, not part of the Dashboard. If you ever feel like converting this feature to a plugin that you’d maintain, we’d gladly link to it 👌

zcallan commented:
Cheers @arturi! I'm glad to have it merged finally :)

You guys have done some great work on this library since I made this PR, keep up the awesome work.

I was thinking of moving the logic of this PR into a small standalone JS library at some point as I've written very similar code to this in a few projects for various things and it's getting tedious to rewrite it.

That could then have an Uppy plugin easily built over it and replaces this PR.

Since the browser history API is pretty concrete at this point, it would mean there'd be very little maintenance to do on the packages which would be great. I just simply haven't had the time yet! I'll try get around to it soon :)
